Floating point random benchmark

NevinBR opened the pull request:
In preparation for #33455, this PR adds benchmarks for generating random floating-point numbers in the interval `0..<1`.

NevinBR committed:
Merge pull request #1 from apple/master

Syncing from upstream

NevinBR committed:
Benchmark for unit interval

NevinBR committed:
Minor cleanup

NevinBR committed:
Copyright date

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

NevinBR committed:
Added single-binade and asymmetric ranges

Also removed unneeded legacy factors from new benchmarks, and reformatted the `RandomValues` array literal to make it more readable.

NevinBR commented on benchmark/single-source/RandomValues.swift:
```diff
@@ -27,6 +27,10 @@ public let RandomValues = [
   BenchmarkInfo(name: "RandomDoubleLCG", runFunction: run_RandomDoubleLCG,
     tags: [.api], legacyFactor: 2),
+  BenchmarkInfo(name: "RandomDoubleUnitDef", runFunction: run_RandomDoubleUnitDef,
+    tags: [.api], legacyFactor: 100),
```
Okay, I’ve removed the legacy factors from the new benchmarks.

stephentyrone commented:
> Okay, I’ve added benchmarks for those ranges.

Thank you!

> I also notice that most of these functions are identical except for the range, the RNG, and the number of iterations. So I’m thinking of making a helper function to avoid the repetition. I’m not super familiar with benchmarking though—is there any reason I shouldn’t do that?

I think that's perfectly reasonable. _Do_ make sure that it doesn't fall off a performance cliff, since we want to be benchmarking the RNG and transform, not the benchmark overhead, but I don't expect that it will.

NevinBR commented:
Does this look like a good approach?

```
@inline(never)
public func run_RandomDoubleUnitDef(_ N: Int) {
  randomDoubleHelperDef(N, 0..<1)
}

@inline(never)
public func run_RandomDoubleBinadeDef(_ N: Int) {
  randomDoubleHelperDef(N, 1..<2)
}

@inline(never)
public func run_RandomDoubleAsymDef(_ N: Int) {
  let range = -(1.0.nextDown) ..< 0.126
  randomDoubleHelperDef(N, range)
}

@inline(__always)
private func randomDoubleHelperDef(
  _ N: Int, _ range: Range<Double>, reps: Int = 1_000
) {
  for _ in 0 ..< N {
    var x = 0.0
    for _ in 0 ..< reps {
      x += Double.random(in: range)
    }
    blackHole(x)
  }
}
```

(And similar for LCG)

stephentyrone commented:
Seems reasonable to me!

NevinBR committed:
Added  helper functions to reduce repetitiveness
